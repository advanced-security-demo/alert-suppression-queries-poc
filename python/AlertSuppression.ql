/**
 * @name Alert suppression
 * @description Generates information about alert suppressions.
 * @kind alert-suppression
 * @id py/alert-suppression
 */

import python

/**
 * An alert suppression comment.
 */
abstract class SuppressionComment extends Comment {
  /** Gets the scope of this suppression. */
  abstract SuppressionScope getScope();

  /** Gets the suppression annotation in this comment. */
  abstract string getAnnotation();

  /**
   * Holds if this comment applies to the range from column `startcolumn` of line `startline`
   * to column `endcolumn` of line `endline` in file `filepath`.
   */
  abstract predicate covers(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  );
}

/**
 * An alert comment that applies to a single line
 */
abstract class LineSuppressionComment extends SuppressionComment {
  LineSuppressionComment() {
    exists(string filepath, int l |
      this.getLocation().hasLocationInfo(filepath, l, _, _, _) and
      any(AstNode a).getLocation().hasLocationInfo(filepath, l, _, _, _)
    )
  }

  /** Gets the scope of this suppression. */
  override SuppressionScope getScope() { result = this }

  override predicate covers(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  ) {
    this.getLocation().hasLocationInfo(filepath, startline, _, endline, endcolumn) and
    startcolumn = 1
  }
}

/**
 * An codeql suppression comment.
 */
class CodeQLSuppressionComment extends LineSuppressionComment {
  string annotation;

  CodeQLSuppressionComment() {
    exists(string all | all = this.getContents() |
      // match `codeql[...]` anywhere in the comment
      annotation =
        all.regexpFind("(?i)\\bcodeql\\s*\\[[^\\]]*\\]", _, _).regexpReplaceAll("^codeql", "lgtm")
      or
      // match `codeql` at the start of the comment and after semicolon
      annotation =
        all.regexpFind("(?i)(?<=^|;)\\s*codeql(?!\\B|\\s*\\[)", _, _)
            .trim()
            .regexpReplaceAll("^codeql", "lgtm")
    )
  }

  /** Gets the suppression annotation in this comment. */
  override string getAnnotation() { result = annotation }
}

/**
 * A noqa suppression comment. Both pylint and pyflakes respect this, so codeql ought to too.
 */
class NoqaSuppressionComment extends LineSuppressionComment {
  NoqaSuppressionComment() { this.getContents().toLowerCase().regexpMatch("\\s*noqa\\s*([^:].*)?") }

  override string getAnnotation() { result = "codeql" }
}

/**
 * The scope of an alert suppression comment.
 */
class SuppressionScope extends @py_comment {
  SuppressionScope() { this instanceof SuppressionComment }

  /**
   * Holds if this element is at the specified location.
   * The location spans column `startcolumn` of line `startline` to
   * column `endcolumn` of line `endline` in file `filepath`.
   * For more information, see
   * [Locations](https://codeql.github.com/docs/writing-codeql-queries/providing-locations-in-codeql-queries/).
   */
  predicate hasLocationInfo(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  ) {
    this.(SuppressionComment).covers(filepath, startline, startcolumn, endline, endcolumn)
  }

  /** Gets a textual representation of this element. */
  string toString() { result = "suppression range" }
}

from SuppressionComment c
select c, // suppression comment
  c.getContents(), // text of suppression comment (excluding delimiters)
  c.getAnnotation(), // text of suppression annotation
  c.getScope() // scope of suppression
