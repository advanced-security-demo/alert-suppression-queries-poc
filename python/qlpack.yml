name: advanced-security-demo/python-alert-suppression
version: 0.0.1
dependencies:
  "codeql/python-all": "0.4.3"
defaultSuite:
  queries: .
