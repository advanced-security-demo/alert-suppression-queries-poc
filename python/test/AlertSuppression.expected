| test.py:4:5:4:12 | Comment # codeql |  codeql | codeql | test.py:4:1:4:12 | suppression range |
| test.py:5:5:5:30 | Comment # codeql[py/line-too-long] |  codeql[py/line-too-long] | codeql[py/line-too-long] | test.py:5:1:5:30 | suppression range |
| test.py:6:5:6:54 | Comment # codeql[py/line-too-long, py/non-callable-called] |  codeql[py/line-too-long, py/non-callable-called] | codeql[py/line-too-long, py/non-callable-called] | test.py:6:1:6:54 | suppression range |
| test.py:7:5:7:27 | Comment # codeql[@tag:security] |  codeql[@tag:security] | codeql[@tag:security] | test.py:7:1:7:27 | suppression range |
| test.py:8:5:8:44 | Comment # codeql[@tag:security,py/line-too-long] |  codeql[@tag:security,py/line-too-long] | codeql[@tag:security,py/line-too-long] | test.py:8:1:8:44 | suppression range |
| test.py:9:5:9:33 | Comment # codeql[@expires:2017-06-11] |  codeql[@expires:2017-06-11] | codeql[@expires:2017-06-11] | test.py:9:1:9:33 | suppression range |
| test.py:10:5:10:70 | Comment # codeql[py/non-callable-called] because I know better than codeql |  codeql[py/non-callable-called] because I know better than codeql | codeql[py/non-callable-called] | test.py:10:1:10:70 | suppression range |
| test.py:11:5:11:23 | Comment # codeql: blah blah |  codeql: blah blah | codeql | test.py:11:1:11:23 | suppression range |
| test.py:12:5:12:37 | Comment # codeql blah blah #falsepositive |  codeql blah blah #falsepositive | codeql | test.py:12:1:12:37 | suppression range |
| test.py:13:5:13:39 | Comment # codeql blah blah -- falsepositive |  codeql blah blah -- falsepositive | codeql | test.py:13:1:13:39 | suppression range |
| test.py:14:5:14:38 | Comment # codeql  [py/non-callable-called] |  codeql  [py/non-callable-called] | codeql  [py/non-callable-called] | test.py:14:1:14:38 | suppression range |
| test.py:15:5:15:14 | Comment # codeql[] |  codeql[] | codeql[] | test.py:15:1:15:14 | suppression range |
| test.py:17:5:17:12 | Comment # codeql |  codeql | codeql | test.py:17:1:17:12 | suppression range |
| test.py:18:5:18:12 | Comment # codeql |  codeql | codeql | test.py:18:1:18:12 | suppression range |
| test.py:19:5:19:34 | Comment # codeql    [py/line-too-long] |  codeql    [py/line-too-long] | codeql    [py/line-too-long] | test.py:19:1:19:34 | suppression range |
| test.py:20:5:20:19 | Comment # codeql codeql |  codeql codeql | codeql | test.py:20:1:20:19 | suppression range |
| test.py:27:13:27:27 | Comment # codeql [func] |  codeql [func] | codeql [func] | test.py:27:1:27:27 | suppression range |
| test.py:29:18:29:38 | Comment # codeql on docstring |  codeql on docstring | codeql | test.py:29:1:29:38 | suppression range |
| test.py:30:15:30:49 | Comment # codeql [py/duplicate-key-in-dict] |  codeql [py/duplicate-key-in-dict] | codeql [py/duplicate-key-in-dict] | test.py:30:1:30:49 | suppression range |
| test.py:36:11:36:24 | Comment # codeql class |  codeql class | codeql | test.py:36:1:36:24 | suppression range |
| test.py:37:22:37:36 | Comment # codeql method |  codeql method | codeql | test.py:37:1:37:36 | suppression range |
| test.py:41:5:41:10 | Comment # noqa |  noqa | codeql | test.py:41:1:41:10 | suppression range |
| test.py:42:5:42:10 | Comment # noqa |  noqa | codeql | test.py:42:1:42:10 | suppression range |
| test.py:47:5:47:32 | Comment # noqa -- Some extra detail. |  noqa -- Some extra detail. | codeql | test.py:47:1:47:32 | suppression range |
| test.py:54:34:54:119 | Comment # noqa: E501; (line too long) pylint: disable=invalid-name; codeql [py/missing-equals] |  noqa: E501; (line too long) pylint: disable=invalid-name; codeql [py/missing-equals] | codeql [py/missing-equals] | test.py:54:1:54:119 | suppression range |
| test.py:58:5:58:70 | Comment # noqa: E501; (line too long) pylint: disable=invalid-name; codeql |  noqa: E501; (line too long) pylint: disable=invalid-name; codeql | codeql | test.py:58:1:58:70 | suppression range |
| test.py:59:5:59:81 | Comment # random nonsense codeql [py/missing-equals] and then some more commentary... |  random nonsense codeql [py/missing-equals] and then some more commentary... | codeql [py/missing-equals] | test.py:59:1:59:81 | suppression range |
| test.py:64:5:64:12 | Comment # codeql |  codeql | codeql | test.py:64:1:64:12 | suppression range |
| test.py:65:5:65:30 | Comment # codeql[py/line-too-long] |  codeql[py/line-too-long] | codeql[py/line-too-long] | test.py:65:1:65:30 | suppression range |
| test.py:71:5:71:65 | Comment # codeql[py/line-too-long] and codeql[py/non-callable-called] |  codeql[py/line-too-long] and codeql[py/non-callable-called] | codeql[py/line-too-long] | test.py:71:1:71:65 | suppression range |
| test.py:71:5:71:65 | Comment # codeql[py/line-too-long] and codeql[py/non-callable-called] |  codeql[py/line-too-long] and codeql[py/non-callable-called] | codeql[py/non-callable-called] | test.py:71:1:71:65 | suppression range |
| test.py:72:5:72:38 | Comment # codeql[py/line-too-long]; codeql |  codeql[py/line-too-long]; codeql | codeql | test.py:72:1:72:38 | suppression range |
| test.py:72:5:72:38 | Comment # codeql[py/line-too-long]; codeql |  codeql[py/line-too-long]; codeql | codeql[py/line-too-long] | test.py:72:1:72:38 | suppression range |
